Source/AppleSocial.swift

```diff
@@ -0,0 +1,119 @@
+//  Copyright © 2020 edX. All rights reserved.
+//
+
+import Foundation
+import UIKit
+import MSAL
+import AuthenticationServices
+
+typealias AppleLoginCompletionHandler =  (OEXRegisteringUserDetails?, _ accessToken: String?, Error?) -> Void
+
+
+class AppleSocial: NSObject {
+
+    static let shared = AppleSocial()
+    var completionHandler: AppleLoginCompletionHandler!
+
+    private override init() {
+        super.init()
+    }
+
+    func loginFromController(controller : UIViewController, completion: @escaping AppleLoginCompletionHandler) {
+
+        completionHandler = completion
+
+        if #available(iOS 13.0, *) {
+            let authorizationProvider = ASAuthorizationAppleIDProvider()
+            let request = authorizationProvider.createRequest()
+            request.requestedScopes = [.fullName, .email]
+
+            let authorizationController = ASAuthorizationController(authorizationRequests: [request])
+            authorizationController.delegate = self
+            authorizationController.presentationContextProvider = self
+            authorizationController.performRequests()
+        }
+        else {
+            let error = NSError(domain: NSCocoaErrorDomain, code: -10000, userInfo: ["messgae": "Apple Sign In not supported prior to iOS 13"])
+            completionHandler(nil, nil, error)
+        }
+    }
+}
+
+@available(iOS 13.0, *)
+extension AppleSocial: ASAuthorizationControllerDelegate {
+
+    func authorizationController(controller: ASAuthorizationController, didCompleteWithError error: Error) {
+        completionHandler(nil, nil, error)
+    }
+
+    func authorizationController(controller: ASAuthorizationController, didCompleteWithAuthorization authorization: ASAuthorization) {
+
+        guard let credentials = authorization.credential as? ASAuthorizationAppleIDCredential else {
+            let error = NSError(domain: NSCocoaErrorDomain, code: ASAuthorizationError.unknown.rawValue, userInfo: ["messgae": "Failed to get credentials"])
+            completionHandler(nil, nil, error)
+            return
+        }
+
+        let firstName = credentials.fullName?.givenName
+        let lastName = credentials.fullName?.familyName
+        let email = credentials.email ?? ""
+
+        let userDetails = OEXRegisteringUserDetails()
+        userDetails.name = "\(firstName ?? "") \(lastName ?? "")"
+        userDetails.email = email
+
+        if let data = credentials.identityToken, let code = String(data: data, encoding: .utf8) {
+            if userDetails.email?.isEmpty ?? true {
+                let tokenDetails = try? decode(jwtToken: code)
+                userDetails.email = tokenDetails?["email"] as? String ?? ""
+            }
+            completionHandler?(userDetails, code, nil)
+        }
+        else {
+            let error = NSError(domain: NSCocoaErrorDomain, code: ASAuthorizationError.failed.rawValue, userInfo: ["messgae": "unable to extract apple identity token"])
```

Comment:
instead of using `NSCocoaErrorDomain` please use class name as domain so callee function can have idea from where error was generated.

instead of using `messgae` as key, please use `NSLocalizedDescriptionKey` so when callee function prints or display information, the value will be availaibe by using `error.localizedDescription`  

```suggestion
            let error = NSError(domain: "AppleSocial", code: ASAuthorizationError.failed.rawValue, userInfo: [NSLocalizedDescriptionKey: "unable to extract apple identity token"])
```